### PR TITLE
Improve Invoke-WebRequest xml and json errors format

### LINKED ISSUE #13251: Invoke-WebRequest: The error message is damaged beyond repair
## Steps to reproduce

```powershell
iwr 'http://github-production-release-asset-2e65be.s3.amazonaws.com/'
```

## Expected behavior

```none
Invoke-WebRequest: Forbidden

Error:
    Code: AccessDenied
    Message: Access Denied
    RequestId: 229E67C9F0B26053
    HostId: kUI8bnH/eZxSAjznVy7+RQpaBv1yHVAP7w9Vh/Z2LYdrrPfn1E8E3y176OHzhKwwdkRMZ0RygbQ=
```

## Actual behavior

```none
Invoke-WebRequest:
AccessDeniedAccess Denied4GEN7KFG5Q8QAT8J/8ZJ9JizWnxWHeo0m3wU/rl7Kh7xzbV89arVlDlaSbYwEGELtFblhWEQAKxeaWmeY61sUti5ISk=
```

## Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.3
PSEdition                      Core
GitCommitId                    7.0.3
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary

Currently Invoke-WebRequest completely mangles XML error responses:

After this PR it recognizes XML and Json and throws better errors:
4x

No changes for text and html errors.

## PR Context

Fixes #13251

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 1]
```diff
@@ -1535,8 +1535,17 @@ protected override void ProcessRecord()
                                     try
                                     {
                                         reader = new StreamReader(StreamHelper.GetResponseStream(response));
-                                        // remove HTML tags making it easier to read
-                                        detailMsg = System.Text.RegularExpressions.Regex.Replace(reader.ReadToEnd(), "<[^>]*>", string.Empty);
+                                        if (ContentHelper.IsXml(contentType))
+                                        {
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1539]: I'd read all here with ReadToEnd() and pass the string to new FormatErrorMessage() where we could process XML, Json, HTML and Text.

**Comment by iSazonov** [comment]
Formally LGTM. It would great to see output examples for all cases (XML, HTML, Json and Text).

**Comment by CarloToso** [comment]
I'm completely satisfied with XML errors
Should I try to remove brackets in JSON errors? (Probably with regex)
Should I try to parse html errors as XML?
How can I distingush Html from text errors?
Where should I move FormatErrorMessage? (StreamHelper.cs?)

**Comment by CarloToso** [comment]
@iSazonov We could stop removing HTML tags (it generates ugly errors with far too much white space in many cases); and format XML without removing tags.

**Comment by iSazonov** [comment]
> @iSazonov We could stop removing HTML tags (it generates ugly errors with far too much white space in many cases); and format XML without removing tags.

Yes, it looks ugly. I don't know why this was done. If there is simple way (like for XML or JSON) to re-format HTML you could add this too.

**Comment by CarloToso** [comment]
Unfortunately I dont think there is an easy way to re-format HTML without using a third party library (ex. AngleSharp).
Where should I move FormatErrorMessage? (StreamHelper.cs? some other place?)

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 2]
```diff
@@ -1927,5 +1928,42 @@ private static StreamContent GetMultipartFileContent(object fieldName, FileInfo
             return result;
         }
         #endregion Helper Methods
+
+        private static string FormatErrorMessage(string error, string contentType)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1932]: We can move the method to the Helpers region.

**Review by iSazonov — COMMENTED** [review]
Please update the PR description with latest screenshots.

**Comment by CarloToso** [comment]
I reverted to the old behaviour for html and plain text.

**Comment by CarloToso** [comment]
I can't figure out why the tests keep failing, any advice? It seems like "I am a teapot!!!" should work as an error

**Comment by iSazonov** [comment]
> I can't figure out why the tests keep failing, any advice? It seems like "I am a teapot!!!" should work as an error

You can run the test locally with Start-PSPester or even manually step by step.

**Comment by CarloToso** [comment]
@iSazonov I fixed the errors, the test helper service returns plain text but says it's json. This resulted in an error when parsing the json. Solved by adding a try-catch

**Comment by iSazonov** [comment]
> @iSazonov I fixed the errors, the test helper service returns plain text but says it's json. This resulted in an error when parsing the json. Solved by adding a try-catch

I'd expect the helper returns right Json so that we could test the formatting code.

**Comment by CarloToso** [comment]
@iSazonov I'll add tests for XML and Json

**Comment by iSazonov** [comment]
> > @iSazonov I fixed the errors, the test helper service returns plain text but says it's json. This resulted in an error when parsing the json. Solved by adding a try-catch
> 
> I'd expect the helper returns right Json so that we could test the formatting code.

@CarloToso Please clarify what the helper service returns and how tests work now. If it returns non-json how we check json scenario?

**Comment by CarloToso** [comment]
@iSazonov The helper service now returns:

1. Plain text, but declares it's application/json (old test, unchanged) -> error caught by try-catch -> old behaviour
2. Json, application/json (new test) -> the text is formatted
3. Xml, application/xml (new test) -> the text is formatted

**Comment by iSazonov** [comment]
@CarloToso Please update the PR description and update to actual screenshots.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 3]
```diff
@@ -1821,6 +1821,69 @@ private static StreamContent GetMultipartFileContent(object fieldName, FileInfo
...
+                        OmitXmlDeclaration = true
+                    };
+
+                    if (doc.FirstChild is XmlDeclaration)
+                    {
+                        XmlDeclaration decl = doc.FirstChild as XmlDeclaration;
+                        settings.Encoding = Encoding.GetEncoding(decl.Encoding);
+                    }
+
+                    StringBuilder stringBuilder = new StringBuilder();
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1849]:
```suggestion
                    StringBuilder stringBuilder = new();
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 4]
```diff
@@ -1821,6 +1821,60 @@ private static StreamContent GetMultipartFileContent(object fieldName, FileInfo
...
+            }
+            catch
+            {
+                // Ignore errors
+            }
+            
+            if (string.IsNullOrEmpty(formattedError))
+            {
+                // Remove HTML tags making it easier to read
+                formattedError = System.Text.RegularExpressions.Regex.Replace(error, "<[^>]*>", string.Empty);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1872]: Should we add Environment.NewLine here too?

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1872]: I'd rather leave the old behaviour.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 5]
```diff
@@ -1821,6 +1821,60 @@ private static StreamContent GetMultipartFileContent(object fieldName, FileInfo
...
+                        XmlDeclaration decl = doc.FirstChild as XmlDeclaration;
+                        settings.Encoding = Encoding.GetEncoding(decl.Encoding);
+                    }
+
+                    StringBuilder stringBuilder = new();
+                    using XmlWriter xmlWriter = XmlWriter.Create(stringBuilder, settings);
+                    doc.Save(xmlWriter);
+                    string xmlString = stringBuilder.ToString();
+
+                    formattedError = Environment.NewLine + xmlString;
```
[the remark this answers is not in the document]
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1853]: We probably could but I'd rather leave it like this, I like the symmetry between
`formattedError = Environment.NewLine + xmlString; and formattedError` = `Environment.NewLine + jsonString;` 😅

**Review by iSazonov — APPROVED** [review]

**Pull request merged**